packages/scheduler/src/Scheduler.js

```diff
@@ -187,6 +187,11 @@ function workLoop(hasTimeRemaining, initialTime) {
       if (typeof continuationCallback === 'function') {
         currentTask.callback = continuationCallback;
         markTaskYield(currentTask, currentTime);
+        if (!hasTimeRemaining || shouldYieldToHost()) {
```

Comment:
As is, we'd end up checking `shouldYieldToHost` twice on every frame.

Reply:
Are you saying you would just always break here so that continuations are always scheduled? If so, we would lose the ability to continue working on a task's continuations for the entire frame until the browser requests a yield, which seems bad.

Note also that on line 124 we check `shouldYieldToHost` so we're already checking it twice per task. This would mean we check it three times per continuation (if it isn't expired) so we could refactor this to remove one check.

Reply:
If we did this then the PR title would be "Continuations should yield before executing".

Reply:
Yeah I’m saying if a continuation is returned, we should assume that `shouldYield` was already checked and it returned `true`. Checking again is redundant.

The check on line 124 is still necessary for yielding between separate tasks — not all tasks use time slicing (i.e. check `shouldYield` and post a continuation). In practice, only React does, and only for rendering (it doesn’t when running passive effects, for example).

The key part of this bug is what happens when an *expired* task posts a continuation. On lines 173-174, we will never exit the work loop if there are expired tasks remaining. We should just remove line 173 entirely IMO. Though that will require updating some tests.

So I would title this PR something more along the lines of “expired tasks should be allowed to yield to the main thread (even though that doesn’t really make sense)”.

And then I think there are at least two follow ups:

- Force `shouldYieldToHost` to return `false` if the task is “expired” or Immediate. It used to do this but the logic was too complicated. We should do something more clever like set the frame deadline to 0.
– Schedule immediate callbacks as microtasks with `queueMicrotask`, instead of adding them to our userspace queue.

Reply:
This makes sense but I have a concern with it.

Line 173 allows expired tasks one opportunity to complete. Without it, expired tasks would start, skip work, and immediately get re-scheduled. If we change `shouldYieldToHost` to take expirations into account again, then this is avoided, but now expired tasks would continue work indefinitely until they're finished, never yielding to the browser.

That's the same bug that I'm attempting to fix here, but without the continuation re-direct.

Is this a valid concern or am I missing something?